ddl/db_partition_test.go

```diff
@@ -761,6 +761,134 @@ func (s *testIntegrationSuite5) TestAlterTableDropPartition(c *C) {
+
+	tk.MustExec("ALTER TABLE e3 EXCHANGE PARTITION p1 WITH TABLE e2;")
+	tk.MustGetErrCode("ALTER TABLE e3 EXCHANGE PARTITION p0 WITH TABLE e2", tmysql.ErrRowDoesNotMatchPartition)
+	tk.MustGetErrCode("ALTER TABLE e3 EXCHANGE PARTITION p2 WITH TABLE e2", tmysql.ErrRowDoesNotMatchPartition)
+	tk.MustGetErrCode("ALTER TABLE e3 EXCHANGE PARTITION p3 WITH TABLE e2", tmysql.ErrRowDoesNotMatchPartition)
+	tk.MustExec("ALTER TABLE e3 EXCHANGE PARTITION p0 with TABLE e2 WITHOUT VALIDATION")
+
+	tk.MustQuery("select * from e3 partition(p0)").Check(testkit.Rows("1", "5"))
+	tk.MustQuery("select * from e2").Check(testkit.Rows())
+
+}
+
+func (s *testIntegrationSuite4) TestExchangePartitionTableCompatiable(c *C) {
+
+	type testCase struct {
+		ptSQL       string
+		ntSQL       string
+		exchangeSQL string
+		err         bool
+	}
+	cases := []testCase{
+		{
+			"create table pt (id int not null) partition by hash (id) partitions 4;",
+			"create table nt (id int(1) not null);",
+			"alter table pt exchange partition p0 with table nt;",
+			false,
+		},
+		{
+			"create table pt1 (id int not null, fname varchar(3)) partition by hash (id) partitions 4;",
+			"create table nt1 (id int not null, fname varchar(4));",
+			"alter table pt1 exchange partition p0 with table nt1;",
+			true,
+		},
+		{
+			"create table pt2 (id int not null, salary decimal) partition by hash(id) partitions 4;",
+			"create table nt2 (id int not null, salary decimal(3,2));",
+			"alter table pt2 exchange partition p0 with table nt2;",
+			true,
+		},
+		{
+			"create table pt3 (id int not null, salary decimal) partition by hash(id) partitions 1;",
+			"create table nt3 (id int not null, salary decimal(10, 1));",
+			"alter table pt3 exchange partition p0 with table nt3",
+			true,
+		},
+		{
+			"create table pt4 (id int not null) partition by hash(id) partitions 1;",
+			"create table nt4 (id1 int not null);",
+			"alter table pt4 exchange partition p0 with table nt4;",
+			true,
+		},
+		{
+			"create table pt5 (id int not null, primary key (id)) partition by hash(id) partitions 1;",
+			"create table nt5 (id int not null);",
+			"alter table pt5 exchange partition p0 with table nt5;",
+			true,
+		},
+		{
+			"create table pt6 (id int not null, salary decimal, index idx (id, salary)) partition by hash(id) partitions 1;",
+			"create table nt6 (id int not null, salary decimal, index idx (salary, id));",
+			"alter table pt6 exchange partition p0 with table nt6;",
+			true,
+		},
+		{
+			"create table pt7 (id int not null, index idx (id) invisible) partition by hash(id) partitions 1;",
+			"create table nt7 (id int not null, index idx (id));",
+			"alter table pt7 exchange partition p0 with table nt7;",
+			false,
+		},
+		{
+			"create table pt8 (id int not null, index idx (id)) partition by hash(id) partitions 1;",
+			"create table nt8 (id int not null, index id_idx (id));",
+			"alter table pt8 exchange partition p0 with table nt8;",
+			true,
+		},
+	}
```

Comment:
Add some test case for:
- generated column (virtual and stored)
- unique index
- foreign key
- auto_increament
- `DEFAULT`
- origin table is `view`

Also, Add some test for these experimental features (need enable a switch in config firstly):
- auto_random
- expression index
- enable `alter-primary-key`

Reply:
Sorry for the late reply. Due to daily work, what is mentioned in the review will be repaired at the weekend or late at night.

Reply:
All that mentioned I have fixed now. @Deardrops PTAL
- For an expression index, I had added more test about it
- For `auto_random`:  I'm not dealing with during exchange partition. Due to meta id of the table has exchanged, the random id will be regenerate when data coming. if check auto_random, it seems that only need to check `AutoRandomBits`, I donot know if there are any other side effects.

Reply:
@tangenta PTAL

Reply:
@zhaox1n Yes, we need to check `AutoRandomBits`. `Exchange partition` is only allowed to execute when `AutoRandomBits` of two tables are equals.